### DISCUSSION TITLE
How to convert the values of the json to string or char array

### DISCUSSION THREAD

**Discussion by Basilisvirus** [nlohmann/json#3741] [Q&A] [answered]
My .json file:

```
{
    "foo": {
        "Too": "1",
        "Moo": "2"
}
}
```

I try to parse/read the Too's and Moo's values and save them in a char or string type variable.

I can successfully print them on the terminal/console using the below code, but I want to convert these values to string, char and put them in a variable, but I get errors 1 and 2:

```
for (auto& [key, value] : membersIterator->items())
{
            std::cout << key << " : " << value << "\n\r";
    //*name = value; //Gives me error 1 when I try to do this conversion
    //strcpy(name, value.c_str()); //Gives me error 2 when I try to use the c_str() function.
}
```

![image](https://user-images.githubusercontent.com/18421898/189531659-a26ec1e4-baea-45e9-934e-a49e467285c6.png)

The errors:

![image](https://user-images.githubusercontent.com/18421898/189531863-e1b80f2c-dc3a-4499-9a5a-0f8a882d450a.png)

So the variables "key" and "value" here are of a certain type of variable created by nlohmann/Json library and I cannot convert it to a char array or string array.

Same thing if I try:

```
for (json::iterator it = membersIterator->begin(); it != membersIterator->end(); ++it)
{
    std::cout << it.key() << " : " << it.value() << "\n\r";
    //*name = it.value();//Gives me error 1 when I try to do this conversion
   // strcpy(name, it.value.c_str()); //Gives me error 2 when I try to use the c_str() function.
}
```
I will get my cout as normal, but I cannot use the c.str() or the conversion to the char array.

**Comment by falbrechtskirchinger** [comment, marked as answer]
All values are instances of `json`(or `basic_json`, really). To convert them to any other type, use:

`get<Type>()`, `get_ref<[const] Type&>()`, `get_to()`, or, if you want to specify a default value `value()`.

Check the documentation:
https://json.nlohmann.me/api/basic_json/get/
https://json.nlohmann.me/api/basic_json/get_ref/
https://json.nlohmann.me/api/basic_json/get_to/
https://json.nlohmann.me/api/basic_json/value/

In your specific case you could use:
```cpp
it.value().get_ref<const std::string&>().c_str();
```

Don't use `strcpy` – ever. Use `strncpy` instead.

**Reply by Basilisvirus** [reply]
Okay, thank you, the documentation helped here. In my case I used a slight variation of your suggestion:

For my first example I can use this:

`std::string jsonValueString = value.get_ref<json::string_t&>();`

For my second example I can use this:

`std::string jsonValueString = it.value().get_ref<json::string_t&>();`


And you meant `strcpy`/`strncpy `instead of `stdcpy`/`stdncpy  `right? :p

**Reply by falbrechtskirchinger** [reply]
> Okay, thank you, the documentation helped here. In my case I used a slight variation of your suggestion:
> 
> For my first example I can use this:
> 
> `std::string jsonValueString = value.get_ref<json::string_t&>();`
>
> For my second example I can use this:
> 
> `std::string jsonValueString = it.value().get_ref<json::string_t&>();`

You could also use:
```cpp
// reference; no copies
const auto &jsonValueString = value.get_ref<const json::string_t&>();
// copy
auto jsonValueString = value.get<json::string_t>();
```
 
> And you meant `strcpy`/`strncpy `instead of `stdcpy`/`stdncpy `right? :p

Duh. Fixed. ;-)
